### ISSUE TITLE
Post method __get extremely slow

### ISSUE THREAD

**Issue by josec89** [jgrossi/corcel#67]
I've been having some troubles with the `__get($key)` method in the `Post` model.

It makes a queries looking for metas every time a key is not found in the *post* model. This is very time consuming and avoids any possible cache (I have a cache and still have like **750** queries to meta files)

![screen shot 2016-02-12 at 01 06 02](https://cloud.githubusercontent.com/assets/860516/12995081/0e028c30-d125-11e5-83cc-309c82308394.png)

I'd suggest another approach, removing the `__get` method and getting all the needed information with *accessors* like done before.

What do you think?

**Comment by kriskbx** [comment]
Corcels queries don't show up in my DebugBar at all. How did you achieved that?

**Comment by josec89** [comment]
I think it's because I am accessing to the new attributes like `title` (instead of `post_title`), I guess the method doesn't found `title` in the attributes and makes the meta() query.

```
        <div class="section-title">
            <h1>{{ $post->title }}</h1>
        </div>
```

I think this place of the code is making the extra queries
```
        if (!isset($this->$key)) {
            if (isset($this->meta()->get()->$key)) {
                return $this->meta()->get()->$key;
            }
        } 
````
Why exactly do we make `meta()->get()` ? The `meta` relationship has been already loaded and thus `meta` (without parenthesis) should be accessible

**Comment by jgrossi** [comment]
I've included a test to check if `$post->title` is getting value from `getTitleAttribute()` or `$post->meta->title` and the result was from `getTitleAttribute()`.

```
PHPUnit 4.2.2 by Sebastian Bergmann.

Configuration read from /home/vagrant/Github/corcel/phpunit.xml

.............................

Time: 977 ms, Memory: 8.00Mb

OK (29 tests, 200 assertions)
```

Any ideias?

**Comment by jgrossi** [comment]
@josec89 The last test shows that where you're using

**Comment by jgrossi** [comment]
@josec89 The last test shows that when you type `$post->title` you get the `post_title` value and not `$post->meta->title` value. Check the commit c02a53ab030dfd622ae5cc6c17e84704eb1eeba6.

And about `$this->meta()->get()->$key` thats the same as `$this->meta->$key` and think this is not responsible for the multiple queries.

The last test I told, for me, proves the code below is working as expected:

```
if (!isset($this->$key)) {
    if (isset($this->meta()->get()->$key)) {
        return $this->meta()->get()->$key;
    }
} 
```

**Comment by josec89** [comment]
Hello,

As far as I understand Eloquent `$this->meta()->get()` creates a query and gets the result, while `$this->meta` doesn't create a query if you added that relationship to the `with` array. Isn't that right?. 

Anyway, I'll investigate further and I'll get to know what's happening! :)

**Comment by jgrossi** [comment]
Monitoring MySQL activity for just an `echo $post->title`:

With `protected $with = ['meta'];` in `Post.php`:

```
2016-02-12T17:48:36.581611Z   16 Connect	homestead@localhost on corcel-dev using TCP/IP
2016-02-12T17:48:36.582228Z   16 Prepare	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:48:36.582692Z   16 Execute	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:48:36.583148Z   16 Close stmt
2016-02-12T17:48:36.583875Z   16 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:48:36.584353Z   16 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:48:36.584882Z   16 Close stmt
2016-02-12T17:48:36.595893Z   16 Prepare	select * from `wp_posts` where `wp_posts`.`ID` = ? limit 1
2016-02-12T17:48:36.596826Z   16 Execute	select * from `wp_posts` where `wp_posts`.`ID` = '1' limit 1
2016-02-12T17:48:36.597544Z   16 Close stmt
2016-02-12T17:48:36.598915Z   16 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in (?)
2016-02-12T17:48:36.599583Z   16 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in ('1')
2016-02-12T17:48:36.600575Z   16 Close stmt
2016-02-12T17:48:36.602038Z   16 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = ? and `wp_postmeta`.`post_id` is not null
2016-02-12T17:48:36.602719Z   16 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = '1' and `wp_postmeta`.`post_id` is not null
2016-02-12T17:48:36.603690Z   16 Close stmt
2016-02-12T17:48:36.606264Z   16 Quit
```

And removing `protected $with = ['meta'];` from `Post.php`:

```
2016-02-12T17:49:13.814637Z   17 Connect	homestead@localhost on corcel-dev using TCP/IP
2016-02-12T17:49:13.815625Z   17 Prepare	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:49:13.816212Z   17 Execute	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:49:13.817010Z   17 Close stmt
2016-02-12T17:49:13.817941Z   17 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:49:13.818638Z   17 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:49:13.819372Z   17 Close stmt
2016-02-12T17:49:13.831048Z   17 Prepare	select * from `wp_posts` where `wp_posts`.`ID` = ? limit 1
2016-02-12T17:49:13.831712Z   17 Execute	select * from `wp_posts` where `wp_posts`.`ID` = '1' limit 1
2016-02-12T17:49:13.832584Z   17 Close stmt
2016-02-12T17:49:13.834548Z   17 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = ? and `wp_postmeta`.`post_id` is not null
2016-02-12T17:49:13.835111Z   17 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = '1' and `wp_postmeta`.`post_id` is not null
2016-02-12T17:49:13.835821Z   17 Close stmt
2016-02-12T17:49:13.837898Z   17 Quit
```

With `with` variable Eloquent is making more queries`. Ideas?

**Comment by josec89** [comment]
Of course, because `with` means, hey! load this relationship as it's important form me. This is cool when you are getting more than one record, because it gets that relationship for all records in only one SQL call.

Can you test changing `$this->meta()->get()->$key` ---> `$this->meta->$key`

**Comment by jgrossi** [comment]
You're right :-)

Results with `$this->meta()->get()->$key`

```
2016-02-12T17:57:08.261091Z   18 Connect	homestead@localhost on corcel-dev using TCP/IP
2016-02-12T17:57:08.261825Z   18 Prepare	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:57:08.262324Z   18 Execute	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:57:08.262778Z   18 Close stmt
2016-02-12T17:57:08.263434Z   18 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:57:08.263958Z   18 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:57:08.264664Z   18 Close stmt
2016-02-12T17:57:08.276701Z   18 Prepare	select * from `wp_posts` where `wp_posts`.`ID` = ? limit 1
2016-02-12T17:57:08.277396Z   18 Execute	select * from `wp_posts` where `wp_posts`.`ID` = '1' limit 1
2016-02-12T17:57:08.278032Z   18 Close stmt
2016-02-12T17:57:08.279341Z   18 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in (?)
2016-02-12T17:57:08.279790Z   18 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in ('1')
2016-02-12T17:57:08.280413Z   18 Close stmt
2016-02-12T17:57:08.281725Z   18 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = ? and `wp_postmeta`.`post_id` is not null
2016-02-12T17:57:08.282207Z   18 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` = '1' and `wp_postmeta`.`post_id` is not null
2016-02-12T17:57:08.282848Z   18 Close stmt
2016-02-12T17:57:08.284636Z   18 Quit
```

Results with `$this->meta->$key`:

```
2016-02-12T17:58:00.213013Z   19 Connect	homestead@localhost on corcel-dev using TCP/IP
2016-02-12T17:58:00.213661Z   19 Prepare	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:58:00.214014Z   19 Execute	set names 'utf8' collate 'utf8_unicode_ci'
2016-02-12T17:58:00.214477Z   19 Close stmt
2016-02-12T17:58:00.215030Z   19 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:58:00.215642Z   19 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` is null and `wp_postmeta`.`post_id` is not null
2016-02-12T17:58:00.216327Z   19 Close stmt
2016-02-12T17:58:00.228369Z   19 Prepare	select * from `wp_posts` where `wp_posts`.`ID` = ? limit 1
2016-02-12T17:58:00.229123Z   19 Execute	select * from `wp_posts` where `wp_posts`.`ID` = '1' limit 1
2016-02-12T17:58:00.229750Z   19 Close stmt
2016-02-12T17:58:00.231104Z   19 Prepare	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in (?)
2016-02-12T17:58:00.231720Z   19 Execute	select * from `wp_postmeta` where `wp_postmeta`.`post_id` in ('1')
2016-02-12T17:58:00.233279Z   19 Close stmt
2016-02-12T17:58:00.236423Z   19 Quit
```

**Comment by josec89** [comment]
Yay! :D

**Comment by jgrossi** [comment]
Fixed #69 :-)

**Comment by yoramdelangen** [comment]
When using `$with` in the model, Eloquent populates the relationship directly when the model is set. If you do `$this->meta()->get()` it queries the results, while `$this->meta->get()` only works if the meta relation is already eager-loaded. In this case it could be more efficient indeed to do it like this. But when the result set is pretty large, Eloquent will fetch all meta records directly into the model and can slow down drastically..

**Comment by josec89** [comment]
I've been doing more research....

My main problem is that I've cached a pack of posts (my pagination) and then I expect 0 queries when I reload, but I was getting the same amount of queries as in the first shot. This happened because the `meta` relationship is not cached. And, for this, I've created an accessor to get all that information:

```
    /**
     * Get all the metas.
     *
     * @return array
     */
    public function getMetasAttribute()
    {
        $metas = [];
        foreach ($this->meta as $meta) {
            $metas[$meta->meta_key] = $meta->value;
        }

        return $metas;
    }
```

Having this, the `__get` method wouldn't work with cached queries :(

I still have more queries than expected but I think I'm getting through this! :D

First shot, no cache
=====
![screen shot 2016-02-14 at 11 47 53](https://cloud.githubusercontent.com/assets/860516/13033273/da5a5ece-d310-11e5-90e6-3b86243d8120.png)

I make 1 query for posts + 1 query for **all** meta... COOL. But then I make 2 queries for every post that have a featured image :-1: I'm working on the way to fix this!

Cached 
======

![screen shot 2016-02-14 at 11 50 30](https://cloud.githubusercontent.com/assets/860516/13033281/2d49d9ca-d311-11e5-8ae0-3098c8762f8a.png)

No more queries! This is what I wanted

**Comment by jgrossi** [comment]
I really don't know, but is there any way to automatically cache relation (`meta`) with Eloquent?

**Comment by josec89** [comment]
It looks like you can't

http://stackoverflow.com/questions/28844611/cache-eloquent-relationship-query

But maybe I should parse the data (get the relationships and stuff) before caching it.

**Comment by jgrossi** [comment]
Yep! I think this is more an application requirement than the tool, right?

**Comment by dark-panda** [comment]
I have had a similar issue where the database was being hit when you would do the following:

```php
$post = Post::query();
```

Which you might want to do start a fresh query chain. You'd end up with queries like this:

```sql
select 
  * 
from 
  `wp_postmeta` 
where 
  `wp_postmeta`.`post_id` is null 
  and `wp_postmeta`.`post_id` is not null
```

That will return an empty set, of course, but it's still unnecessary to hit the database when you know it shouldn't be returning anything, as is the case where you have a new instance of `Post` and no `ID` to go along with it.

I've created two patches with potential solutions to the issue and since they might intersect with some of the issues here I figured I'd post the patches here rather than start a new issue ticket.

* https://github.com/dark-panda/corcel/tree/dont-fetch-meta-on-null-with-pass-through - in this case, if we don't have an `ID` yet, we just pass the `__get` up to `parent`. 

* https://github.com/dark-panda/corcel/tree/dont-fetch-meta-on-null-with-set-new-meta-collection - in this case, we instantiate a new `PostMetaCollection` and assign it to the `Post` to mimic what we would have received from the database if we did have an `ID`.

I'm more partial to the second solution, but neither of them are perfect. I just figured I'd post them here for discussion to see if anyone had a better solution;
